### ISSUE TITLE
Did not wait for the output

### ISSUE THREAD

**Issue by jzyxn** [XinyuanWangCS/PromptAgent#7]
![Quicker_20240512_151305](https://github.com/XinyuanWangCS/PromptAgent/assets/76219656/b2bb1df9-3011-4cc6-ac89-3b086209fbdd)

I used TinyLlama-1.1B-Chat-v1.0 for test(dataset temporal_sequences only have 16 items after deleting), but I didn't see the output after running 30 minutes and the one iteration hadn't completed.

Part of the process is as follows：
![Quicker_20240512_152130](https://github.com/XinyuanWangCS/PromptAgent/assets/76219656/c4ac6fe5-6abe-4266-82fb-1bd84b75fb91)

**Comment by XinyuanWangCS** [comment]
If I understand correct, this is because the method of PromptAgent uses error examples to do self-reflection. If there is no error,  the "while loop" doesn't stop. Later, I will add similar prompt to learn from correct examples. This may solve your issue, but it will be different from the original method in the paper. Also, I'd like to add that PromptAgent or our prompt optimization method is designed for a little bit harder tasks (base model makes errors). If the base model has already been able to solve the task with a simple prompt, then it is not necessary to spend money to optimize prompt for that.

**Comment by jzyxn** [comment]
Okay, I probably get it.

**Comment by CSJDeveloper** [comment]
> Okay, I probably get it.

I do not think this algorithm can work because the code is totally different from the one reported in the ICLR paper. The main reason is that once there is a problem, the tree structure should be very simple and it should converge even faster. 

__If the task is simpler, the convergence speed should be faster instead of blocking there (as you reported) without doing anything.__

Therefore, I have checked the code, and I personally do not agree with the author.  

So, try other things, and we should wait until the authors make more updates to make this project __at least run smoothly__. 

It is a bit sad, but that's the way it is in the research world, nowadays; stories outweigh implementation.

**Comment by XinyuanWangCS** [comment]
Thanks for pointing out the issue. I agree with you for simpler task the convergence speed is faster, and it actually is, since the paths will be early stopped if the prompt 0 is very good and new prompts are no better, so the searching tree will be shallower. But there are two things to clarify or discuss. First, what this all-correct "blocks" is self-reflection step, which tries to extract information from data points. It is not MCTS algorithm's problem. Second, we have considered the case when this kind of all correct batch happens, and we will resample batches. But I appologize that we didn't consider what if it's all correct in the whole training set. All correct in the training set means a really good prompt (or the task is solved by the model). It will be a question whether we should even do the searching if the state 0 has already reached the maximum reward, like the robot has already been put at the goal point, should it take any actions then. I can certainly break the loop as you said, making the algorithm converge faster, but if so, the prompt willl not be updated (no new prompts). What the user may expect may be seeing some new prompts, so I am considering adding more to the information extraction part. This will be different from the paper, but make the code more robust for users in the application perspective. Sorry for letting you guys meet these issues. But I think our code is following the method in the paper, not "totally different". I have graduated, so the update of the code is a bit slow, but please bring up any issue you meet, which I will fix later. Hope this can make you feel better about research world.

**Comment by jzyxn** [comment]
I thank you both for your academic rigour and look forward to seeing a more robust code.

**Comment by CSJDeveloper** [comment]
> Thanks for pointing out the issue. I agree with you for simpler task the convergence speed is faster, and it actually is, since the paths will be early stopped if the prompt 0 is very good and new prompts are no better, so the searching tree will be shallower. But there are two things to clarify or discuss. First, what this all-correct "blocks" is self-reflection step, which tries to extract information from data points. It is not MCTS algorithm's problem. Second, we have considered the case when this kind of all correct batch happens, and we will resample batches. But I appologize that we didn't consider what if it's all correct in the whole training set. All correct in the training set means a really good prompt (or the task is solved by the model). It will be a question whether we should even do the searching if the state 0 has already reached the maximum reward, like the robot has already been put at the goal point, should it take any actions then. I can certainly break the loop as you said, making the algorithm converge faster, but if so, the prompt willl not be updated (no new prompts). What the user may expect may be seeing some new prompts, so I am considering adding more to the information extraction part. This will be different from the paper, but make the code more robust for users in the application perspective. Sorry for letting you guys meet these issues. But I think our code is following the method in the paper, not "totally different". I have graduated, so the update of the code is a bit slow, but please bring up any issue you meet, which I will fix later. Hope this can make you feel better about research world.

Thanks a lot for the further reply. More discussion would always be better to gain more insights. Allowing LLMs to self-check and then back to revise the reasoning is a core direction in this domain, at least from my perspective. I track many important work, such as Boosting of Thughts, "Large language models cannot self-correct reasoning yet" and your PromptAgent. For some of them, I am still waiting for the code to be released. Anyway, please let me know when the code is totally ready! Thanks! Because I hope this direction at least makes sense.

**Comment by XinyuanWangCS** [comment]
> ![Quicker_20240512_151305](https://private-user-images.githubusercontent.com/76219656/329832247-b2bb1df9-3011-4cc6-ac89-3b086209fbdd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NTA4ODMsIm5iZiI6MTcxNTk1MDU4MywicGF0aCI6Ii83NjIxOTY1Ni8zMjk4MzIyNDctYjJiYjFkZjktMzAxMS00Y2M2LWFjODktM2IwODYyMDlmYmRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDEyNTYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0NGIxNGRkNGZjNDliYmFiNGYwYjM4NGIxNGU2NmFhYWY0ZWM0NDdiYzhhZTIxNzQxM2VkZjQ0MzIwNDZkNGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pjPQHjEZap4GYHRDzAh5JCwq8Wwo_wcfJNjA-qf_fzU)
> 
> I used TinyLlama-1.1B-Chat-v1.0 for test(dataset temporal_sequences only have 16 items after deleting), but I didn't see the output after running 30 minutes and the one iteration hadn't completed.
> 
> Part of the process is as follows： ![Quicker_20240512_152130](https://private-user-images.githubusercontent.com/76219656/329832653-c4ac6fe5-6abe-4266-82fb-1bd84b75fb91.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NTA4ODMsIm5iZiI6MTcxNTk1MDU4MywicGF0aCI6Ii83NjIxOTY1Ni8zMjk4MzI2NTMtYzRhYzZmZTUtNmFiZS00MjY2LTgyZmItMWJkODRiNzVmYjkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDEyNTYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmZWE1NTQzYTIwMTk2M2MwMTE3NmUzZTYxM2YwNTk2YWM1MGQ3NWY1OWIyOTkxMzE4ZTAwNDQ2YjliMDhlMmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XvlRuuFS2nTF3I3hOz1tQZISjfMppWX2SvB8CjR4kcI)

I have updated the code. Right now the program won't be stuck because of all correct issue. You can change the arguments in the example_config.yaml file to use your base model like TinyLlama. I had a simple try of TinyLlama, and I don't think this small model is able to understand the long prompt. Sometimes it generates nothing, sometimes repeated sentences, sometimes it generates an answer but now in a correct format <answer>(X)</answer>, so the answer still can't be extracted from response. I use some general arguments (max_new_tokens, repetition_penalty) for the huggingface text generation model, you may check and modify them or add new arguments in the hf_textgeneration_model.py file in the language_model folder. I would recommand you using instruction-tuned llama7b or mistral7b as base model, as we said expert-level prompts are for advanced models. Also, the optim model should be a strong model like GPT-3.5.

**Comment by jzyxn** [comment]
> > ![Quicker_20240512_151305](https://private-user-images.githubusercontent.com/76219656/329832247-b2bb1df9-3011-4cc6-ac89-3b086209fbdd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NTA4ODMsIm5iZiI6MTcxNTk1MDU4MywicGF0aCI6Ii83NjIxOTY1Ni8zMjk4MzIyNDctYjJiYjFkZjktMzAxMS00Y2M2LWFjODktM2IwODYyMDlmYmRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDEyNTYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0NGIxNGRkNGZjNDliYmFiNGYwYjM4NGIxNGU2NmFhYWY0ZWM0NDdiYzhhZTIxNzQxM2VkZjQ0MzIwNDZkNGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pjPQHjEZap4GYHRDzAh5JCwq8Wwo_wcfJNjA-qf_fzU)
> > I used TinyLlama-1.1B-Chat-v1.0 for test(dataset temporal_sequences only have 16 items after deleting), but I didn't see the output after running 30 minutes and the one iteration hadn't completed.
> > Part of the process is as follows： ![Quicker_20240512_152130](https://private-user-images.githubusercontent.com/76219656/329832653-c4ac6fe5-6abe-4266-82fb-1bd84b75fb91.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NTA4ODMsIm5iZiI6MTcxNTk1MDU4MywicGF0aCI6Ii83NjIxOTY1Ni8zMjk4MzI2NTMtYzRhYzZmZTUtNmFiZS00MjY2LTgyZmItMWJkODRiNzVmYjkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDEyNTYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmZWE1NTQzYTIwMTk2M2MwMTE3NmUzZTYxM2YwNTk2YWM1MGQ3NWY1OWIyOTkxMzE4ZTAwNDQ2YjliMDhlMmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XvlRuuFS2nTF3I3hOz1tQZISjfMppWX2SvB8CjR4kcI)
> 
> I have updated the code. Right now the program won't be stuck because of all correct issue. You can change the arguments in the example_config.yaml file to use your base model like TinyLlama. I had a simple try of TinyLlama, and I don't think this small model is able to understand the long prompt. Sometimes it generates nothing, sometimes repeated sentences, sometimes it generates an answer but now in a correct format (X), so the answer still can't be extracted from response. I use some general arguments (max_new_tokens, repetition_penalty) for the huggingface text generation model, you may check and modify them or add new arguments in the hf_textgeneration_model.py file in the language_model folder. I would recommand you using instruction-tuned llama7b or mistral7b as base model, as we said expert-level prompts are for advanced models. Also, the optim model should be a strong model like GPT-3.5.

Thanks again for the code changes and your discussion with another friend.

**Comment by CSJDeveloper** [comment]
> > ![Quicker_20240512_151305](https://private-user-images.githubusercontent.com/76219656/329832247-b2bb1df9-3011-4cc6-ac89-3b086209fbdd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NTA4ODMsIm5iZiI6MTcxNTk1MDU4MywicGF0aCI6Ii83NjIxOTY1Ni8zMjk4MzIyNDctYjJiYjFkZjktMzAxMS00Y2M2LWFjODktM2IwODYyMDlmYmRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDEyNTYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0NGIxNGRkNGZjNDliYmFiNGYwYjM4NGIxNGU2NmFhYWY0ZWM0NDdiYzhhZTIxNzQxM2VkZjQ0MzIwNDZkNGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pjPQHjEZap4GYHRDzAh5JCwq8Wwo_wcfJNjA-qf_fzU)
> > I used TinyLlama-1.1B-Chat-v1.0 for test(dataset temporal_sequences only have 16 items after deleting), but I didn't see the output after running 30 minutes and the one iteration hadn't completed.
> > Part of the process is as follows： ![Quicker_20240512_152130](https://private-user-images.githubusercontent.com/76219656/329832653-c4ac6fe5-6abe-4266-82fb-1bd84b75fb91.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NTA4ODMsIm5iZiI6MTcxNTk1MDU4MywicGF0aCI6Ii83NjIxOTY1Ni8zMjk4MzI2NTMtYzRhYzZmZTUtNmFiZS00MjY2LTgyZmItMWJkODRiNzVmYjkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDEyNTYyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmZWE1NTQzYTIwMTk2M2MwMTE3NmUzZTYxM2YwNTk2YWM1MGQ3NWY1OWIyOTkxMzE4ZTAwNDQ2YjliMDhlMmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XvlRuuFS2nTF3I3hOz1tQZISjfMppWX2SvB8CjR4kcI)
> 
> I have updated the code. Right now the program won't be stuck because of all correct issue. You can change the arguments in the example_config.yaml file to use your base model like TinyLlama. I had a simple try of TinyLlama, and I don't think this small model is able to understand the long prompt. Sometimes it generates nothing, sometimes repeated sentences, sometimes it generates an answer but now in a correct format (X), so the answer still can't be extracted from response. I use some general arguments (max_new_tokens, repetition_penalty) for the huggingface text generation model, you may check and modify them or add new arguments in the hf_textgeneration_model.py file in the language_model folder. I would recommand you using instruction-tuned llama7b or mistral7b as base model, as we said expert-level prompts are for advanced models. Also, the optim model should be a strong model like GPT-3.5.

Thanks! Days later, I will try to reproduce Table 5 and Figure 5 of the paper. From my perspective, the way to self-improve with LLMs is insane, as LLMs frequently output hallucinations (especially small-scale ones, llama7b), and there should be complex mechanisms to avoid this before making it work. This may be a good try. Discuss with you then. 

Thanks again.
